### ISSUE TITLE
Feature Request: MAL/AniList Sync and Integration

### ISSUE THREAD

**Issue by hotsno** [animeloyalty/animeloyalty#12]
I just wanna start off by saying thank you so much for making this. It's really incredible what you've done so far.

I totally understand if you're uninterested in adding this, but many people, myself included, use MyAnimeList / AniList or some other tracking service to keep their watchlists up to date. I use it so I know which episode I left off on across devices. Adding a feature that would automatically update the episode number on these trackers as the episode reaches a certain % watch completion would be amazing.

This feature can be done in a _really_ simple way. If you've used Tachiyomi (Android manga app), what they do is they just take the title of the manga, use the AniList / MAL API to search the name of the manga, and then display the results to the user. It allows the user to change the search term in case the title being used is far off from the title in the tracker. The user chooses the correct entry, and then that entry is associated with that manga. Then, once the user completes a chapter, the tracker is updated to match what chapter they just read. While I'm at it, I'd like to suggest a way to show which episodes have been watched (maybe change the episode thumbnail to have a check mark over it). This could be extended into the tracker integration so when a tracker is added where your episode watch count > 0, they'll automatically be marked as watched in the app as well.

A more advanced feature would be the ability to import your watchlists from these trackers, and have them displayed in a "Watchlist" section. This would obviously require a ton more work since the app doesn't even have its own "watchlist" feature yet.

Personally, I've used the AniList API and have some experience with it, and know it's super easy to use and works great. I've heard mixed things about the MAL API. Unfortunately I don't know Electron and am not fluent in JavaScript so I don't know if my efforts would be worthwhile to try and add this stuff myself.

**Comment by hotsno** [comment]
Here are some other repos that do similar things.
Tachiyomi: Android manga reader which has tracker support - https://github.com/tachiyomiorg/tachiyomi
MAL-Sync: Browser extension that updates MAL/AL after detecting episode completion on various sites (including CR and Funi) - https://github.com/tachiyomiorg/tachiyomi

**Comment by hotsno** [comment]
Wow I just read https://github.com/animeloyalty/animeloyalty/issues/5 and I feel so dumb now, though I'm assuming the "tracking addition" will just locally track, not using 3rd party tracking sites. If you could add AniList and MAL integration into this, it'd be a godsend, though I understand if that's not what you want :)

**Comment by Deathspike** [comment]
I like this feature, but I'm a little concerned about the technical difficulties. But first things first:

You're 100% correct in your assumption. The personal edition with the library feature will download a copy of the provider metadata (e.g. Crunchyroll) so it becomes available offline, and will then allow you to mark episodes and track watching progress in that metadata. It's straightforward and simple. There are no external dependencies of any kind. This is a work in progress, albeit progress has been a bit slow lately (because life).

Now with that out of the way, the suggestion to manually pair an entry with an series is in theory pretty straight forward. But it becomes a little more complicated once you realize that episodes don't necessarily have numbers, aren't necessarily in the series to begin with, aren't always numbered correctly, and season-spanning series aren't considered one series. So let's break this down:

1. I add https://www.crunchyroll.com/my-hero-academia and pair it with https://myanimelist.net/anime/31964/Boku_no_Hero_Academia. Now the episodes 1-13 will be able to match and be marked correctly. Yay!
2. I start watching Season 2, episode 14 (because 13.5 is a recap). But My Hero Academia only has 13 episodes according to MAL. So instead of pairing a series on CR to a series on MAL, we have to pair a season on CR to a series on MAL. A little awkward, but do-able. Let's assume we paired Season 2 to https://myanimelist.net/anime/33486/Boku_no_Hero_Academia_2nd_Season.
3. Now I start watching Season 2, episode 14 again. So I mark episode 14 of Season 2 watched on MAL. And I keep going until episode 38. You might see the problem here... there are only 25 episodes on MAL and the moment I watch the first episode of Season 2, I've marked 14 individual episodes as watched on MAL. There's no bullet-proof solution AFAIK.
    * If I recount the episodes, then episode 13.5 becomes episode 1. So skipping the recap (which is a different 'series' on MAL), I end up watching the second episode and I'll keep going to a total of 26 episodes... rather than 25.
    * If I subtract the minimum episode number from the total, skipping those with 0.5, then I would end up with (14-13) for the first episode. This is the correct solution, until we encounter something like half-season uploads as is the case for https://www.crunchyroll.com/one-piece.
4. Let's ignore the situation at (3) and watch 13.5 for now, the recap special. This episode can't be matched to anything in Season 2 either, but is instead https://myanimelist.net/anime/35262/Boku_no_Hero_Academia_2nd_Season__Hero_Note. So... instead of matching seasons on CR to series on MAL I should... match episodes... per series? This is getting cumbersome.
5. Ok, screw it. Let's watch https://www.crunchyroll.com/attack-on-titan instead. Wait, what the hell is the 'PV' doing there?!

Let me reiterate, I **do like the feature**. But I have no idea how to handle the many, many edge-cases that are so prevalent in anime. I might be overthinking this whole thing, though. Suggestions and feedback is most welcome on this!

**Comment by hotsno** [comment]
Yeah, I agree, there do seem to be some edge cases and this makes it more difficult than it needs to be. From what I see, one doable way would be to as you said, be able to pair different MAL/AL entries with different seasons, since that's how most shows are organized on MAL. Though, correct me if I'm wrong, but it seems to me that Crunchyroll and MAL use the same organization. For shows that are organized by season in CR, they have entries by season in MAL (ex. MHA is by season in both, Hunter x Hunter is all one entry in both). With that, the main issue that remains are the specials that screw up the episode count. The obvious thing I can think of is to add an "offset" option that the user can adjust for these cases.

Also, I wanted to mention this again: I was suggesting that all of the "setting up" of the tracking is done manually by the user (like in Tachiyomi). This basically ensures that the app is never at fault for messing up tracking, and allows the user to change stuff if there's a mistake.

Another thing to consider is if this really needs to be "perfect." Personally I'd still love to have this feature even if it didn't work perfectly for every single anime.

Finally, I wanted to share this: https://github.com/MALSync/MAL-Sync-Backup
It's MAL-Sync's database which has Crunchyroll episode links mapped to MAL IDs. I don't know if you'd want to use it, but from my experience their data has been very accurate.

**Comment by Deathspike** [comment]
Unfortunately, even seasons don't map 1:1. For example:

Re:ZERO -Starting Life in Another World- Season 2
https://www.crunchyroll.com/rezero-starting-life-in-another-world-

There are 25 episodes in Season 2, but it's a split-cour season. 
It aired in Summer 2020, went on hiatus, and then resumed in Winter 2021.
And that means there are two entries on MAL:

* https://myanimelist.net/anime/39587/re_zero_kara_hajimeru_isekai_seikatsu_2nd_season
* https://myanimelist.net/anime/42203/Re_Zero_kara_Hajimeru_Isekai_Seikatsu_2nd_Season_Part_2

---

A possible solution would be allow multi-series mapping to single provider season.
Then all you'd need is the possibility to mark episodes as exclusions from tracking (specials, etc).
It would work like this (continuing to use https://www.crunchyroll.com/rezero-starting-life-in-another-world- as example):

* There's a mapping button in the season header.
    * You click the button to open the mapping screen.
* You can add one or more external series here.
    * You add https://myanimelist.net/anime/39587/re_zero_kara_hajimeru_isekai_seikatsu_2nd_season
    * You add https://myanimelist.net/anime/42203/Re_Zero_kara_Hajimeru_Isekai_Seikatsu_2nd_Season_Part_2
* A preview will be shown. Each episode matches with a series title and its respective episode number.
    * Counting will be based on **episode order**, not number/title. So...
    * Episode 26 matches Re:Zero kara Hajimeru Isekai Seikatsu 2nd Season, episode 1.
    * Episode 27 matches Re:Zero kara Hajimeru Isekai Seikatsu 2nd Season, episode 2.
    * Episode 39 matches Re:Zero kara Hajimeru Isekai Seikatsu 2nd Season, episode 13.
    * Episode 40 matches Re:Zero kara Hajimeru Isekai Seikatsu 2nd Season Part 2, episode 1.
* If there's something incorrectly, you have two options:
    * You can mark an episode as "don't count". Useful for specials like 13.5, you can exclude them from mapping.
    * You can add an offset to a series. This is useful in case a series isn't complete on the provider.

Things intentionally neglected in this design:
* Specials: They aren't worth the effort, and generally aren't interesting to list anyway.
* Movies: Some may be mappable if they have their own season, but some will not have that.

What do you think of this solution?
(On the topic of MAL Sync, I prefer not to use external databases but a live-searchable API instead)

**Comment by hotsno** [comment]
I think this solution works great! Let me know if you ever need me to test something/an opinion on something (hotsno#0001 on Discord if you use that). Looking forward to seeing your next versions/editions.
